File: problems/0301-remove-invalid-parentheses/solution.cpp
// Link: https://leetcode.cn/problems/remove-invalid-parentheses/
// Level: Hard
// Date: 2025-05-12

class Solution {
private:
    bool isValid(const string& s) {
        int count = 0;
        for (char c : s) {
            if (c == '(') {
                count++;
            } else if (c == ')') {
                if (count == 0) {
                    return false;
                }
                count--;
            }
        }
        return count == 0;
    }

public:
    vector<string> removeInvalidParentheses(string s) {
        vector<string> res;
        queue<string> q;
        unordered_set<string> visited;
        q.push(s);
        visited.insert(s);
        bool found = false;


        while (!q.empty()) {
            int levelSize = q.size();
            unordered_set<string> currentLevel;
            
            for (int i = 0; i < levelSize; i++) {
                string current = q.front();
                q.pop();

                if (isValid(current)) {
                    res.push_back(current);
                    found = true;
                }

                if (!found) {
                    for (int j = 0; j < current.size(); j++) {
                        if (current[j] != '(' && current[j] != ')') {
                            continue;
                        }
                        string newStr = current.substr(0, j) + current.substr(j + 1);
                        if (visited.find(newStr) == visited.end()) {
                            currentLevel.insert(newStr);
                            visited.insert(newStr);
                        }
                    }
                }
            }

            if (found) {
                break;
            }
            for (const string& str : currentLevel) {
                q.push(str);
            }
        }
        return res;
    }
};

File: problems/0301-remove-invalid-parentheses/solution.go
// Link: https://leetcode.cn/problems/remove-invalid-parentheses/
// Level: Hard
// Date: 2025-05-12

func removeInvalidParentheses(s string) []string {
	res := []string{}
	queue := []string{s}
	visited := make(map[string]bool)
	visited[s] = true
	found := false

	for len(queue) > 0 {
		levelSize := len(queue)
		curLevel := make(map[string]bool)

		for i := 0; i < levelSize; i++ {
			cur := queue[0]
			queue = queue[1:]

			if isValid(cur) {
				res = append(res, cur)
				found = true
			}
			if !found {
				for j := 0; j < len(cur); j++ {
					if cur[j] != '(' && cur[j] != ')' {
						continue
					}
					newStr := cur[:j] + cur[j+1:]
					if !visited[newStr] {
						curLevel[newStr] = true
						visited[newStr] = true
					}
				}
			}
		}
		if found {
			break
		}
		for str := range curLevel {
			queue = append(queue, str)
		}
	}
	return res
}

func isValid(s string) bool {
	count := 0
	for _, c := range s {
		if c == '(' {
			count++
		} else if c == ')' {
			count--
			if count < 0 {
				return false
			}
		}
	}
	return count == 0
}
